In [1]:
# Udacity Machine Learning Nano Degree
# Capstone Project
# Walmart Trip Type Classification Kaggle Dataset
# Jeremy Jesse - 2016-06-26

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

dataset_train = pd.read_csv("train.csv")


# This is a sample of what the data looks like
print dataset_train.head()

#This is the number of rows
num_rows=dataset_train.shape[0]
print "This is the number of rows."
print num_rows
#This is the number of columns
num_col=dataset_train.shape[1]
print "This is the number of columns."
print num_col








   TripType  VisitNumber Weekday           Upc  ScanCount  \
0       999            5  Friday  6.811315e+10         -1   
1        30            7  Friday  6.053882e+10          1   
2        30            7  Friday  7.410811e+09          1   
3        26            8  Friday  2.238404e+09          2   
4        26            8  Friday  2.006614e+09          2   

   DepartmentDescription  FinelineNumber  
0     FINANCIAL SERVICES          1000.0  
1                  SHOES          8931.0  
2          PERSONAL CARE          4504.0  
3  PAINT AND ACCESSORIES          3565.0  
4  PAINT AND ACCESSORIES          1017.0  
This is the number of rows.
647054
This is the number of columns.
7


We have a dataset with 7 columns wide and 647,054 rows long.  It consists of transaction data from Walmart.
TripType is a classification of the trip as determined by Walmart.
VisitNumber is a unique identifier for a specific receipt at a cash register.
Weekday is simply the day of the week.
UPC is the universal product code.
ScanCount is the number of the given UPC that was purchased.  If the value is negative then there was a return.
DepartmentDescription is the name of the department, which is a logical grouping of similar UPC's.
FinelineNumber is a more detailed grouping of UPC's than DepartmentDescription.

The count of different variables is not the same.  There must be some missing data.  The difference in count between these variables is quite small (647054-642925)/647054=0.638%.  Later any rows with missing data will be removed

In [3]:
#http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.describe.html
#https://stackoverflow.com/questions/22235245/calculate-summary-statistics-of-columns-in-dataframe/22235393
dataset_train.describe().transpose()

count          mean           std    min           25%  \
TripType        647054.0  5.858451e+01  1.576355e+02    3.0  2.700000e+01   
VisitNumber     647054.0  9.616764e+04  5.554549e+04    5.0  4.926800e+04   
Upc             642925.0  3.060698e+10  9.120134e+10  834.0  3.400001e+09   
ScanCount       647054.0  1.108878e+00  7.007762e-01  -12.0  1.000000e+00   
FinelineNumber  642925.0  3.726885e+03  2.780966e+03    0.0  1.404000e+03   

                         50%           75%           max  
TripType        3.900000e+01  4.000000e+01  9.990000e+02  
VisitNumber     9.707400e+04  1.443160e+05  1.913470e+05  
Upc             7.050103e+09  3.006531e+10  9.789707e+11  
ScanCount       1.000000e+00  1.000000e+00  7.100000e+01  
FinelineNumber  3.352000e+03  5.501000e+03  9.998000e+03

In [4]:
dataset_train.count()

TripType                 647054
VisitNumber              647054
Weekday                  647054
Upc                      642925
ScanCount                647054
DepartmentDescription    645693
FinelineNumber           642925
dtype: int64

The various trip types and department descriptions are seen below.

It can be seen that for a given trip type it is possible to see obvious meaning as to what the trip represents..  The table below shows the number of trips which fall in a given trip type for each department.

Trip type 3 has a very large number of visits in Financial Services, while there aren't very many visits for the other departments.  This is the financial services trip.

Trip type 5 must be primarily for those who are ill.  Both Pharmacy OTC and Pharmacy RX have have a large number of visits.

In [5]:
#http://pandas.pydata.org/pandas-docs/stable/groupby.html
#http://pandas.pydata.org/pandas-docs/stable/generated/pandas.pivot_table.html#pandas.pivot_table
#https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.count_nonzero.html#numpy.count_nonzero
#
table = pd.pivot_table(dataset_train, values='VisitNumber', index=['DepartmentDescription'], columns=['TripType'], aggfunc=np.count_nonzero)
table

TripType                       3      4       5       6       7       8    \
DepartmentDescription                                                       
1-HR PHOTO                     2.0    NaN     1.0     1.0     3.0     NaN   
ACCESSORIES                    2.0    1.0    16.0     3.0    11.0     6.0   
AUTOMOTIVE                    20.0    3.0    40.0     5.0    31.0    56.0   
BAKERY                         5.0   13.0    66.0    23.0   897.0   472.0   
BATH AND SHOWER                8.0    NaN    25.0     6.0     5.0    36.0   
BEAUTY                        24.0    7.0   232.0    23.0    75.0  1112.0   
BEDDING                        4.0    NaN     7.0     2.0     8.0     6.0   
BOOKS AND MAGAZINES            4.0    NaN    15.0     2.0    15.0    17.0   
BOYS WEAR                      9.0    NaN    15.0     1.0     6.0    16.0   
BRAS & SHAPEWEAR               NaN    NaN     9.0     NaN     2.0     6.0   
CAMERAS AND SUPPLIES           5.0    NaN     2.0     NaN     NaN     NaN   
CANDY, TOBACCO, COOKIES       24.0    9.0   133.0   674.0   151.0   579.0   
CELEBRATION                   17.0    3.0    83.0    30.0    44.0   101.0   
COMM BREAD                     4.0    8.0    94.0    32.0  1341.0   476.0   
CONCEPT STORES                 2.0    NaN     NaN     NaN     NaN     NaN   
COOK AND DINE                 12.0    1.0    54.0    14.0    62.0    47.0   
DAIRY                         13.0   11.0   237.0    92.0  1273.0  1813.0   
DSD GROCERY                   59.0   73.0   614.0   360.0  2097.0  4747.0   
ELECTRONICS                   15.0    NaN    18.0     2.0    12.0    11.0   
FABRICS AND CRAFTS             7.0    NaN    61.0    11.0    26.0    38.0   
FINANCIAL SERVICES          5369.0    5.0    50.0    14.0    65.0     NaN   
FROZEN FOODS                   7.0    6.0   114.0    84.0  2278.0   433.0   
FURNITURE                      1.0    NaN     NaN     NaN     1.0     NaN   
GIRLS WEAR, 4-6X  AND 7-14     2.0    1.0    13.0     1.0    12.0    16.0   
GROCERY DRY GOODS             11.0   18.0   384.0   154.0  3354.0  1532.0   
HARDWARE                      11.0    1.0    46.0     8.0    37.0    54.0   
HEALTH AND BEAUTY AIDS         NaN    NaN     NaN     NaN     NaN     NaN   
HOME DECOR                     2.0    2.0    37.0     5.0    24.0    22.0   
HOME MANAGEMENT               14.0    3.0    64.0    11.0    24.0    59.0   
HORTICULTURE AND ACCESS        NaN    4.0    16.0     4.0    13.0    10.0   
...                            ...    ...     ...     ...     ...     ...   
LARGE HOUSEHOLD GOODS         76.0    NaN     NaN     NaN     NaN     NaN   
LAWN AND GARDEN                9.0    NaN    60.0    14.0    54.0    24.0   
LIQUOR,WINE,BEER               9.0    NaN    36.0  1000.0    71.0    49.0   
MEAT - FRESH & FROZEN          1.0    1.0    21.0    25.0  1527.0    24.0   
MEDIA AND GAMING              10.0    NaN    15.0     3.0     7.0     5.0   
MENS WEAR                     37.0    7.0    77.0     9.0    46.0    33.0   
MENSWEAR                       3.0    NaN     1.0     NaN     NaN     1.0   
OFFICE SUPPLIES               12.0    2.0    83.0    14.0    41.0    84.0   
OPTICAL - FRAMES               1.0    NaN   396.0     NaN     3.0     3.0   
OPTICAL - LENSES               NaN    NaN    85.0     NaN     NaN     NaN   
OTHER DEPARTMENTS              NaN    NaN     NaN     1.0     1.0     NaN   
PAINT AND ACCESSORIES          3.0    1.0    10.0     1.0    12.0    24.0   
PERSONAL CARE                 40.0   32.0   922.0    73.0   206.0  2650.0   
PETS AND SUPPLIES             11.0    2.0    95.0    26.0    45.0   552.0   
PHARMACY OTC                  45.0  563.0  5281.0    51.0   159.0   648.0   
PHARMACY RX                    NaN    4.0  2714.0     NaN     2.0    10.0   
PLAYERS AND ELECTRONICS        1.0    NaN     NaN     NaN     3.0     1.0   
PLUS AND MATERNITY             1.0    NaN     1.0     NaN     2.0     2.0   
PRE PACKED DELI                NaN    2.0    25.0    25.0   826.0    61.0

Let us look a little closer to find out how many unique values there are for each column in the dataframe.
There are 38 trip types, where 999 stands for other.
There were 95,674 unique visitors.  What this really means is unique receipts.  Them same person could have shopped at Walamrt more than once.
Unsurprisingly, there were 7 weekdays.  These text values will need to be changed to numbers to work with many classification algorithms.
The scancount or number of a given items purchased are all reasonable.
There are 69 departments.  They all seem reasonable as groupings of items.
There are 5,196 finelines.  Having worked at Walmart, I know that finelines are controlled by individual buyers, who all have their own way of placing UPC's within them.  Therefore Finelines wouldn't show useful information.  Also, there are simply too many of them to create that many dummy variables.  My computer would run out of RAM.



In [6]:
#http://chrisalbon.com/python/pandas_list_unique_values_in_column.html
print 'The number of trip types are'
print len(dataset_train.TripType.unique())
print 'The unique values are'
print dataset_train.TripType.unique()
print ''
print 'The number of VisitNumbers are'
print len(dataset_train.VisitNumber.unique())
print 'The unique values are'
print dataset_train.VisitNumber.unique()
print ''
print 'The number of Weekdays are'
print len(dataset_train.Weekday.unique())
print 'The unique values are'
print dataset_train.Weekday.unique()

# There are too many UPC's in the dataset to perform this calculation.  The kearnel dies.
# Although UPC numbers wouldn't be useful anyway.
#print 'The number of Upcs are'
#print len(dataset_train.Upc.unique())
#print 'The unique values are'
#print dataset_train.Upc.unique()

print ''
print 'The number of ScanCount are'
print len(dataset_train.ScanCount.unique())
print 'The unique values are'
print dataset_train.ScanCount.unique()

print ''


print 'The number of DepartmentDescription are'
print len(dataset_train.DepartmentDescription.unique())
print 'The unique values are'
print dataset_train.DepartmentDescription.unique()

print''
print 'The number of FinelineNumber are'
print len(dataset_train.FinelineNumber.unique())
print 'The unique values are'
print dataset_train.FinelineNumber.unique()




The number of trip types are
38
The unique values are
[999  30  26   8  35  41  21   6  42   7   9  39  25  38  15  36  20  37
  32  40   5   3   4  24  33  43  31  27  34  18  29  44  19  23  22  28
  14  12]

The number of VisitNumbers are
95674
The unique values are
[     5      7      8 ..., 191345 191346 191347]

The number of Weekdays are
7
The unique values are
['Friday' 'Saturday' 'Sunday' 'Monday' 'Tuesday' 'Wednesday' 'Thursday']

The number of ScanCount are
39
The unique values are
[ -1   1   2   3   5   6   4  14  -2   9   7  10   8  -3  -5  11  16  -4
  13  15  30  -6  12  20 -12  19  46  23  -7  22  25  24  31  -9  51  17
  18  71 -10]

The number of DepartmentDescription are
69
The unique values are
['FINANCIAL SERVICES' 'SHOES' 'PERSONAL CARE' 'PAINT AND ACCESSORIES'
 'DSD GROCERY' 'MEAT - FRESH & FROZEN' 'DAIRY' 'PETS AND SUPPLIES'
 'HOUSEHOLD CHEMICALS/SUPP' nan 'IMPULSE MERCHANDISE' 'PRODUCE'
 'CANDY, TOBACCO, COOKIES' 'GROCERY DRY GOODS' 'BOYS WEAR'
 'FABRICS AND CR

The correlation map from the customer segments project doesn't help very much.

In [ ]:
#https://github.com/LateralAnalytics/Customer-Segments
import seaborn as sns
corr = dataset_train.corr()
mask = np.zeros_like(corr)
mask[np.triu_indices_from(mask)] = True
with sns.axes_style("white"):
    ax = sns.heatmap(corr, mask=mask, square=True, annot=True, cmap='RdBu')

Now it is time to look at some plots of the data.

A description of each trip is found in "Trip Names.xlsx".

In [ ]:
#https://stackoverflow.com/questions/332289/how-do-you-change-the-size-of-figures-drawn-with-matplotlib
plt.figure(figsize=(15,10))
TripTypeNbr999 = dataset_train[dataset_train.TripType == 999]
TripTypeNbr999Depts = TripTypeNbr999[["TripType","DepartmentDescription"]]
TripTypeNbr999Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 999", rot=90, color="red")
plt.ylabel("Department Count")

In [ ]:
#https://stackoverflow.com/questions/332289/how-do-you-change-the-size-of-figures-drawn-with-matplotlib
plt.figure(figsize=(15,10))
TripTypeNbr999 = dataset_train[dataset_train.TripType == 999]
TripTypeNbr999Depts = TripTypeNbr999[["TripType","DepartmentDescription"]]
TripTypeNbr999Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 999", rot=90, color="red")



plt.figure(figsize=(15,10))
TripTypeNbr30 = dataset_train[dataset_train.TripType == 30]
TripTypeNbr30Depts = TripTypeNbr30[["TripType","DepartmentDescription"]]
TripTypeNbr30Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 30",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr8 = dataset_train[dataset_train.TripType == 8]
TripTypeNbr8Depts = TripTypeNbr8[["TripType","DepartmentDescription"]]
TripTypeNbr8Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 8",rot=90, color="red")


plt.figure(figsize=(15,10))
TripTypeNbr35 = dataset_train[dataset_train.TripType == 35]
TripTypeNbr35Depts = TripTypeNbr35[["TripType","DepartmentDescription"]]
TripTypeNbr35Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 35",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr41 = dataset_train[dataset_train.TripType == 41]
TripTypeNbr41Depts = TripTypeNbr41[["TripType","DepartmentDescription"]]
TripTypeNbr41Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 41",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr21 = dataset_train[dataset_train.TripType == 21]
TripTypeNbr21Depts = TripTypeNbr21[["TripType","DepartmentDescription"]]
TripTypeNbr21Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 21",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr6 = dataset_train[dataset_train.TripType == 6]
TripTypeNbr6Depts = TripTypeNbr6[["TripType","DepartmentDescription"]]
TripTypeNbr6Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 6",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr42 = dataset_train[dataset_train.TripType == 42]
TripTypeNbr42Depts = TripTypeNbr42[["TripType","DepartmentDescription"]]
TripTypeNbr42Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 42",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr7 = dataset_train[dataset_train.TripType == 7]
TripTypeNbr7Depts = TripTypeNbr7[["TripType","DepartmentDescription"]]
TripTypeNbr7Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 7",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr9 = dataset_train[dataset_train.TripType == 9]
TripTypeNbr9Depts = TripTypeNbr9[["TripType","DepartmentDescription"]]
TripTypeNbr9Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 9",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr39 = dataset_train[dataset_train.TripType == 39]
TripTypeNbr39Depts = TripTypeNbr39[["TripType","DepartmentDescription"]]
TripTypeNbr39Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 39",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr25 = dataset_train[dataset_train.TripType == 25]
TripTypeNbr25Depts = TripTypeNbr25[["TripType","DepartmentDescription"]]
TripTypeNbr25Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 25",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr38 = dataset_train[dataset_train.TripType == 38]
TripTypeNbr38Depts = TripTypeNbr38[["TripType","DepartmentDescription"]]
TripTypeNbr38Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 38",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr15 = dataset_train[dataset_train.TripType == 15]
TripTypeNbr15Depts = TripTypeNbr15[["TripType","DepartmentDescription"]]
TripTypeNbr15Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 15",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr36 = dataset_train[dataset_train.TripType == 36]
TripTypeNbr36Depts = TripTypeNbr36[["TripType","DepartmentDescription"]]
TripTypeNbr36Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 36",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr20 = dataset_train[dataset_train.TripType == 20]
TripTypeNbr20Depts = TripTypeNbr20[["TripType","DepartmentDescription"]]
TripTypeNbr20Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 20",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr37 = dataset_train[dataset_train.TripType == 37]
TripTypeNbr37Depts = TripTypeNbr37[["TripType","DepartmentDescription"]]
TripTypeNbr37Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 37",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr32 = dataset_train[dataset_train.TripType == 32]
TripTypeNbr32Depts = TripTypeNbr32[["TripType","DepartmentDescription"]]
TripTypeNbr32Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 32",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr40 = dataset_train[dataset_train.TripType == 40]
TripTypeNbr40Depts = TripTypeNbr40[["TripType","DepartmentDescription"]]
TripTypeNbr40Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 40",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr5 = dataset_train[dataset_train.TripType == 5]
TripTypeNbr5Depts = TripTypeNbr5[["TripType","DepartmentDescription"]]
TripTypeNbr5Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 5",rot=90, color="red")
plt.ylabel("Department Count")

plt.figure(figsize=(15,10))
TripTypeNbr3 = dataset_train[dataset_train.TripType == 3]
TripTypeNbr3Depts = TripTypeNbr3[["TripType","DepartmentDescription"]]
TripTypeNbr3Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 3",rot=90, color="red")
plt.ylabel("Department Count")

plt.figure(figsize=(15,10))
TripTypeNbr4 = dataset_train[dataset_train.TripType == 4]
TripTypeNbr4Depts = TripTypeNbr4[["TripType","DepartmentDescription"]]
TripTypeNbr4Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 4",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr24 = dataset_train[dataset_train.TripType == 24]
TripTypeNbr24Depts = TripTypeNbr24[["TripType","DepartmentDescription"]]
TripTypeNbr24Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 24",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr33 = dataset_train[dataset_train.TripType == 33]
TripTypeNbr33Depts = TripTypeNbr33[["TripType","DepartmentDescription"]]
TripTypeNbr33Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 33",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr43 = dataset_train[dataset_train.TripType == 43]
TripTypeNbr43Depts = TripTypeNbr43[["TripType","DepartmentDescription"]]
TripTypeNbr43Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 43",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr31 = dataset_train[dataset_train.TripType == 31]
TripTypeNbr31Depts = TripTypeNbr31[["TripType","DepartmentDescription"]]
TripTypeNbr31Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 31",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr27 = dataset_train[dataset_train.TripType == 27]
TripTypeNbr27Depts = TripTypeNbr27[["TripType","DepartmentDescription"]]
TripTypeNbr27Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 27",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr34 = dataset_train[dataset_train.TripType == 34]
TripTypeNbr34Depts = TripTypeNbr34[["TripType","DepartmentDescription"]]
TripTypeNbr34Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 34",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr18 = dataset_train[dataset_train.TripType == 18]
TripTypeNbr18Depts = TripTypeNbr18[["TripType","DepartmentDescription"]]
TripTypeNbr18Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 18",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr29 = dataset_train[dataset_train.TripType == 29]
TripTypeNbr29Depts = TripTypeNbr29[["TripType","DepartmentDescription"]]
TripTypeNbr29Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 29",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr44 = dataset_train[dataset_train.TripType == 44]
TripTypeNbr44Depts = TripTypeNbr44[["TripType","DepartmentDescription"]]
TripTypeNbr44Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 44",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr19 = dataset_train[dataset_train.TripType == 19]
TripTypeNbr19Depts = TripTypeNbr19[["TripType","DepartmentDescription"]]
TripTypeNbr19Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 19",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr23 = dataset_train[dataset_train.TripType == 23]
TripTypeNbr23Depts = TripTypeNbr23[["TripType","DepartmentDescription"]]
TripTypeNbr23Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 23",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr22 = dataset_train[dataset_train.TripType == 22]
TripTypeNbr22Depts = TripTypeNbr22[["TripType","DepartmentDescription"]]
TripTypeNbr22Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 22",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr28 = dataset_train[dataset_train.TripType == 28]
TripTypeNbr28Depts = TripTypeNbr28[["TripType","DepartmentDescription"]]
TripTypeNbr28Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 28",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr14 = dataset_train[dataset_train.TripType == 14]
TripTypeNbr14Depts = TripTypeNbr14[["TripType","DepartmentDescription"]]
TripTypeNbr14Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 14",rot=90, color="red")

plt.figure(figsize=(15,10))
TripTypeNbr12 = dataset_train[dataset_train.TripType == 12]
TripTypeNbr12Depts = TripTypeNbr12[["TripType","DepartmentDescription"]]
TripTypeNbr12Depts.DepartmentDescription.value_counts().plot(kind="bar", title="TripType Nbr 12",rot=90, color="red")







Also, outliers will not be removed.  We should not remove outliers just because they are significantly different from most data points.  In retail there are many differnt types of customers.  A particular trip may seem like an outlier compared to the overall data, but it is quite normal for that trip.  Walmart has already grouped each visit_number into the appropriate TripType.

It's already been established that missing values are a small part of this dataset, so they will now be removed.

In [7]:
#http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dropna.html
#https://stackoverflow.com/questions/13413590/how-to-drop-rows-of-pandas-dataframe-whose-value-of-certain-column-is-nan
#Drop rows that have one or more NAN values in them.
dataset_train=dataset_train.dropna()
#Check to validate
dataset_train.describe().transpose()




count          mean           std    min           25%  \
TripType        642925.0  5.833723e+01  1.565599e+02    3.0  2.700000e+01   
VisitNumber     642925.0  9.619987e+04  5.555284e+04    5.0  4.930000e+04   
Upc             642925.0  3.060698e+10  9.120134e+10  834.0  3.400001e+09   
ScanCount       642925.0  1.110203e+00  7.012401e-01  -12.0  1.000000e+00   
FinelineNumber  642925.0  3.726885e+03  2.780966e+03    0.0  1.404000e+03   

                         50%           75%           max  
TripType        3.900000e+01  4.000000e+01  9.990000e+02  
VisitNumber     9.711600e+04  1.443530e+05  1.913470e+05  
Upc             7.050103e+09  3.006531e+10  9.789707e+11  
ScanCount       1.000000e+00  1.000000e+00  7.100000e+01  
FinelineNumber  3.352000e+03  5.501000e+03  9.998000e+03

It's time to create features and adjust the data into a form that most machine learning algorithms use.  In short, we know things about the data intuitively that we need to help the machine learning algorithm see.  Also, we need to change it from long to wide format.

First, Change the the Weekday column to numbers.  Then create a new column which denotes whether or not the visit occured during the weekend.  Then denote whether or not a return occured during the trip.  Turn the DepartmentDescription column into dummy variables which hold the amount spent in that department.

The aggregate the data together so that each VisitNumber only shows up on one row.

In [ ]:
      
#Turn the "Weekday" column into numbers, so they work with algorithms better.
dataset_train.replace('Sunday', 1, inplace=True)
dataset_train.replace('Monday', 2, inplace=True)
dataset_train.replace('Tuesday', 3, inplace=True)
dataset_train.replace('Wednesday', 4, inplace=True)
dataset_train.replace('Thursday', 5, inplace=True)
dataset_train.replace('Friday', 6, inplace=True)
dataset_train.replace('Saturday', 7, inplace=True)

#Create a Weekday/Weekend feature
#http://chrisalbon.com/python/pandas_replace_values.html
dataset_train.loc[dataset_train.Weekday == 1, 'Weekpart'] = 1
dataset_train.loc[dataset_train.Weekday == 2, 'Weekpart'] = 0
dataset_train.loc[dataset_train.Weekday == 3, 'Weekpart'] = 0
dataset_train.loc[dataset_train.Weekday == 4, 'Weekpart'] = 0
dataset_train.loc[dataset_train.Weekday == 5, 'Weekpart'] = 0
dataset_train.loc[dataset_train.Weekday == 6, 'Weekpart'] = 0
dataset_train.loc[dataset_train.Weekday == 7, 'Weekpart'] = 1

#Create a returned Item feature
#https://stackoverflow.com/questions/16327055/how-to-add-an-empty-column-to-a-dataframe
dataset_train['ReturnedItem'] = 0
dataset_train.loc[dataset_train.ScanCount < 0, 'ReturnedItem'] = 1

#Create dummy variables for department description.
#It holds information about what product people want for a certain trip type, but isn't too sparse like finelines.
#http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html
dummy = pd.get_dummies(dataset_train.DepartmentDescription)
dummy_columns=dummy.columns
dataset_train[dummy_columns] = dummy 
dummy_data = dataset_train.iloc[:,9:77]
dummy_data = dummy_data.apply(lambda y: dataset_train["ScanCount"]*y)
dataset_train = dataset_train[['VisitNumber', 'TripType', 'Weekday', 'Weekpart', 'ScanCount', 'ReturnedItem','Upc']]

dataset_train = pd.concat([dataset_train, dummy_data], axis=1)

#Aggregate the data together in a way that is useful for algorithms
#Change long data into wide data
#http://pandas.pydata.org/pandas-docs/stable/groupby.html
grouped = dataset_train.groupby('VisitNumber')
grouped = grouped.aggregate({
'Weekday': np.mean,
'Weekpart': np.mean,
'ScanCount': np.sum,
'ReturnedItem': np.max,
'Upc': np.count_nonzero,
'1-HR PHOTO': np.sum,
'ACCESSORIES': np.sum,
'AUTOMOTIVE': np.sum,
'BAKERY': np.sum,
'BATH AND SHOWER': np.sum,
'BEAUTY': np.sum,
'BEDDING': np.sum,
'BOOKS AND MAGAZINES': np.sum,
'BOYS WEAR': np.sum,
'BRAS & SHAPEWEAR': np.sum,
'CAMERAS AND SUPPLIES': np.sum,
'CANDY, TOBACCO, COOKIES': np.sum,
'CELEBRATION': np.sum,
'COMM BREAD': np.sum,
'CONCEPT STORES': np.sum,
'COOK AND DINE': np.sum,
'DAIRY': np.sum,
'DSD GROCERY': np.sum,
'ELECTRONICS': np.sum,
'FABRICS AND CRAFTS': np.sum,
'FINANCIAL SERVICES': np.sum,
'FROZEN FOODS': np.sum,
'FURNITURE': np.sum,
'GIRLS WEAR, 4-6X  AND 7-14': np.sum,
'GROCERY DRY GOODS': np.sum,
'HARDWARE': np.sum,
'HEALTH AND BEAUTY AIDS': np.sum,
'HOME DECOR': np.sum,
'HOME MANAGEMENT': np.sum,
'HORTICULTURE AND ACCESS': np.sum,
'HOUSEHOLD CHEMICALS/SUPP': np.sum,
'HOUSEHOLD PAPER GOODS': np.sum,
'IMPULSE MERCHANDISE': np.sum,
'INFANT APPAREL': np.sum,
'INFANT CONSUMABLE HARDLINES': np.sum,
'JEWELRY AND SUNGLASSES': np.sum,
'LADIES SOCKS': np.sum,
'LADIESWEAR': np.sum,
'LARGE HOUSEHOLD GOODS': np.sum,
'LAWN AND GARDEN': np.sum,
'LIQUOR,WINE,BEER': np.sum,
'MEAT - FRESH & FROZEN': np.sum,
'MEDIA AND GAMING': np.sum,
'MENS WEAR': np.sum,
'MENSWEAR': np.sum,
'OFFICE SUPPLIES': np.sum,
'OPTICAL - FRAMES': np.sum,
'OPTICAL - LENSES': np.sum,
'OTHER DEPARTMENTS': np.sum,
'PAINT AND ACCESSORIES': np.sum,
'PERSONAL CARE': np.sum,
'PETS AND SUPPLIES': np.sum,
'PHARMACY OTC': np.sum,
'PHARMACY RX': np.sum,
'PLAYERS AND ELECTRONICS': np.sum,
'PLUS AND MATERNITY': np.sum,
'PRE PACKED DELI': np.sum,
'PRODUCE': np.sum,
'SEAFOOD': np.sum,
'SEASONAL': np.sum,
'SERVICE DELI': np.sum,
'SHEER HOSIERY': np.sum,
'SHOES': np.sum,
'SLEEPWEAR/FOUNDATIONS': np.sum,
'SPORTING GOODS': np.sum,
'SWIMWEAR/OUTERWEAR': np.sum,
'TOYS': np.sum,
'WIRELESS': np.sum,
'TripType': np.mean
})







In [ ]:
#grouped.to_csv("grouped.csv", sep='\t')
#grouped.to_pickle("grouped")
import pickle
file_Name = "grouped_pickle"
fileObject = open(file_Name,'wb')
pickle.dump(grouped,fileObject)
fileObject.close()
